brave/src/test/java/com/linecorp/armeria/it/brave/ArmeriaHttpTracingIntegrationTest.java

```diff
@@ -0,0 +1,146 @@
+
+    @Override
+    protected Builder tracingBuilder(Sampler sampler) {
+        return super.tracingBuilder(sampler)
+                    .currentTraceContext(currentTraceContext);
+    }
+
+    @Override
+    protected void init() throws Exception {
+        final ServerBuilder sb = new ServerBuilder();
+        sb.service("/", (ctx, req) -> {
+            if (req.method() == HttpMethod.OPTIONS) {
+                return HttpResponse.of(OK, MediaType.PLAIN_TEXT_UTF_8, "");
+            }
+            return HttpResponse.of(HttpStatus.NOT_FOUND);
+        });
+        sb.service("/foo", (ctx, req) -> HttpResponse.of(OK, MediaType.PLAIN_TEXT_UTF_8, "bar"));
+        sb.service("/extra",
+                   (ctx, req) -> HttpResponse.of(OK, MediaType.PLAIN_TEXT_UTF_8, req.headers().get(EXTRA_KEY)));
+        sb.service("/badrequest", (ctx, req) -> HttpResponse.of(BAD_REQUEST));
+        sb.service("/child", (ctx, req) -> {
+            Tracing.currentTracer().nextSpan().name("child").start().finish();
+            return HttpResponse.of(OK, MediaType.PLAIN_TEXT_UTF_8, "happy");
+        });
+        sb.service("/exception", (ctx, req) -> {
+            throw new Exception();
+        });
+        sb.decorator(BraveService.newDecorator(httpTracing));
+
+        server = sb.build();
+        server.start().join();
+    }
+
+    /**
+     * Postpone url to response.
+     */
+    @Override
+    @Test
+    public void supportsPortableCustomization() throws Exception {
+        httpTracing = httpTracing.toBuilder().serverParser(new HttpServerParser() {
+            @Override
+            public <T> void request(HttpAdapter<T, ?> adapter, T req, SpanCustomizer customizer) {
+                customizer.tag("context.visible", String.valueOf(currentTraceContext.get() != null));
+                customizer.tag("request_customizer.is_span", String.valueOf(customizer instanceof brave.Span));
+            }
+
+            @Override
+            public <T> void response(HttpAdapter<?, T> adapter, T res, Throwable error,
+                                     SpanCustomizer customizer) {
+                super.response(adapter, res, error, customizer);
+                customizer.tag("response_customizer.is_span", String.valueOf(customizer instanceof brave.Span));
+                customizer.tag("http.url", ((ArmeriaHttpServerAdapter) adapter).url(((RequestLog) res)));
+            }
+        }).build();
+        init();
+
+        final String uri = "/foo?z=2&yAA=1";
+        get(uri);
+
+        final Span span = takeSpan();
+        assertThat(span.tags())
+                .containsEntry("http.url", url(uri))
+                .containsEntry("context.visible", "true")
+                .containsEntry("request_customizer.is_span", "false")
+                .containsEntry("response_customizer.is_span", "false");
+    }
+
+    @Override
+    public void notFound() {
+        throw new AssumptionViolatedException("Armeria 404 not supported.");
```

Comment:
please cite TODO of the issue about this as 404 should be traceable I think

Reply:
There's a limitation in Armeria that you cannot decorate a non-existing path mapping. You can bind a service that returns 404 at / and decorate it as a workaround, though.